File: java/house/src/main/java/House.java
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class House {
    private static final String[] SUBJECTS = new String[] {
            "house that Jack built",
            "malt",
            "rat",
            "cat",
            "dog",
            "cow with the crumpled horn",
            "maiden all forlorn",
            "man all tattered and torn",
            "priest all shaven and shorn",
            "rooster that crowed in the morn",
            "farmer sowing his corn",
            "horse and the hound and the horn",
    };
    private static final String[] ACTIONS = new String[] {
            "lay in",
            "ate",
            "killed",
            "worried",
            "tossed",
            "milked",
            "kissed",
            "married",
            "woke",
            "kept",
            "belonged to",
    };

    String verse(int num) {
        StringBuilder sb = new StringBuilder("This is the ").append(SUBJECTS[num - 1]);
        for (int i = num - 2; i >= 0; --i) {
            sb.append(" that ").append(ACTIONS[i]).append(" the ").append(SUBJECTS[i]);
        }
        return sb.append(".").toString();
    }

    String verses(int start, int endInclusive) {
        return IntStream.rangeClosed(start, endInclusive)
                        .mapToObj(this::verse)
                        .collect(Collectors.joining("\n"));
    }

    String sing() {
        return verses(1, 12);
    }
}

File: javascript/house/house.js
export class House {
    constructor() {
        this.entities = [
            'malt',
            'rat',
            'cat',
            'dog',
            'cow with the crumpled horn',
            'maiden all forlorn',
            'man all tattered and torn',
            'priest all shaven and shorn',
            'rooster that crowed in the morn',
            'farmer sowing his corn',
            'horse and the hound and the horn',
        ];
        this.actions = [
            'lay in',
            'ate',
            'killed',
            'worried',
            'tossed',
            'milked',
            'kissed',
            'married',
            'woke',
            'kept',
            'belonged to',
        ];
    }

    static verse(n) {
        const house = new House();
        let output = 'This is the ';
        for (let i = n - 2; i >= 0; i--) {
            output = `${output}${house.entities[i]}\nthat ${house.actions[i]} the `;
        }
        output = `${output}house that Jack built.`;
        return output.split('\n');
    }

    static verses(first, last) {
        let output = [];
        for (let i = first; i <= last; i++) {
            output = output.concat(House.verse(i), ['']);
        }
        output.pop();
        return output;
    }
}
